cpp/src/umap/knn_graph/algo.cuh

```diff
@@ -15,44 +15,118 @@
-                     IntType k, cudaStream_t s)
- */
-template <typename T>
-void launcher(float *X, int x_n, float *X_query, int x_q_n, int d,
-              int64_t *knn_indices, T *knn_dists, int n_neighbors,
-              UMAPParams *params, std::shared_ptr<deviceAllocator> d_alloc,
+template <typename value_idx = int64_t, typename value_t = float,
+          typename umap_inputs>
+void launcher(const umap_inputs &inputsA, const umap_inputs &inputsB,
+              ML::knn_graph<value_idx, value_t> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
+              cudaStream_t stream);
+
+// Instantiation for dense inputs, int64_t indices
+template <>
+void launcher(const ML::manifold_dense_inputs_t<float> &inputsA,
+              const ML::manifold_dense_inputs_t<float> &inputsB,
+              ML::knn_graph<int64_t, float> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
               cudaStream_t stream) {
   std::vector<float *> ptrs(1);
   std::vector<int> sizes(1);
-  ptrs[0] = X;
-  sizes[0] = x_n;
+  ptrs[0] = inputsA.X;
+  sizes[0] = inputsA.n;
+
+  MLCommon::Selection::brute_force_knn(
+    ptrs, sizes, inputsA.d, inputsB.X, inputsB.n, out.knn_indices,
+    out.knn_dists, n_neighbors, d_alloc, stream);
+}
 
-  MLCommon::Selection::brute_force_knn(ptrs, sizes, d, X_query, x_q_n,
-                                       knn_indices, knn_dists, n_neighbors,
-                                       d_alloc, stream);
+// Instantiation for dense inputs, int indices
+template <>
+void launcher(const ML::manifold_dense_inputs_t<float> &inputsA,
+              const ML::manifold_dense_inputs_t<float> &inputsB,
+              ML::knn_graph<int, float> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
+              cudaStream_t stream) {
+  throw raft::exception("Dense KNN doesn't yet support 32-bit integer indices");
 }
+
+template <>
+void launcher(const ML::manifold_sparse_inputs_t<int, float> &inputsA,
+              const ML::manifold_sparse_inputs_t<int, float> &inputsB,
+              ML::knn_graph<int, float> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
+              cudaStream_t stream) {
+  // TODO: Use handle as input here and remove manual creation
+  cusparseHandle_t cusparseHandle;
+  CUSPARSE_CHECK(cusparseCreate(&cusparseHandle));
+
+  MLCommon::Sparse::Selection::brute_force_knn(
+    inputsA.indptr, inputsA.indices, inputsA.data, inputsA.nnz, inputsA.n,
+    inputsA.d, inputsB.indptr, inputsB.indices, inputsB.data, inputsB.nnz,
+    inputsB.n, inputsB.d, out.knn_indices, out.knn_dists, n_neighbors,
+    cusparseHandle, d_alloc, stream, 2 << 14, 2 << 14,
```

Comment:
why 2<<14, specifically?

Reply:
2^14. Just a reasonable starting point to saturate the GPU initially. There's ultimately going to be index_batch_size * query_batch_size number of vector mutiplies and once the GPU is saturated, I didn't find any performance boost with larger batch sizes. Since it's assumed that UMAP will generally be dealing in higher dimensions, I don't want to set it too high, whilst the density of the arrays could end up blowing the GPU memory. If it becomes a problem, we can expose this as an additional argument or compute it based on the density / distribution of the sparse array.